Allow all providers in tag mapping screen

lpichler opened the pull request:
This PR along with the [core PR](https://github.com/ManageIQ/manageiq/pull/20289) enables the following new functions:
- Allows mapping a label to existing managed tags categories
- Allows mapping labels of all provider types  and resource types to managed tag categories
- As a side effect of the above, it allows referencing mapped managed tag categories and entries in policies, rate assignments, reports, expressions, filters, etc.

The representation of this concept is presented externally in the UI and the supporting code as **All Entities**. Internally, `All Entities` is implemented as a tag category without the `<provider-type>:<resource-type>:` prefix, which is what effectively opens mapping up to regular managed tag categories.
![image](https://user-images.githubusercontent.com/76593/93347771-76350d00-f803-11ea-86d1-1d7bf37bb848.png)

Opening up label mapping to managed tag categories presents a number of challenges and edge cases that add some new behavior. One of which is having to support both single-value and multi-value categories. We think we've identified and addresses all edge cases and have implemented tests in both PRs to cover them.

## UI
![bB70F7KfBe](https://user-images.githubusercontent.com/14937244/94283238-0d5f3a80-ff51-11ea-81d8-05b9e3fefc0f.gif)


## Summarization of behavior

### Specific Entity (Provider,... ) - tags/classifications contain prefixes

#### Used Values in UI form:
- Entity: `Instance(Amazon)`
- Resource Label: `environment`
- Tag Category:  `NewCategory` (this category does not have to exist)

Actions:
- create:
  - Tag Mapping:
```
ContainerLabelTagMapping.last
=>

#<ContainerLabelTagMapping:0x00007fb3377cc1c0> {
                       :id => 9,
    :labeled_resource_type => "Vm",
               :label_name => "environment",
              :label_value => nil,
                   :tag_id => 215,
               :created_at => Thu, 20 Aug 2020 14:45:37 UTC +00:00,
               :updated_at => Thu, 20 Aug 2020 14:45:37 UTC +00:00
}
```
Tag created with mapping:

```
ContainerLabelTagMapping.last.tag
=>
#<Tag:0x00007fb31d1e4740> {
      :id => 215,
    :name => "/managed/amazon:vm:newcategory"
}
```
and classification

```
ContainerLabelTagMapping.last.tag.classification
=>
#<Classification:0x00007fb32e6d3330> {
              :id => 215,
     :description => "amazon:vm|NewCategory",
            :icon => nil,
       :read_only => true,
          :syntax => "string",
    :single_value => true,
    :example_text => nil,
          :tag_id => 215,
       :parent_id => nil,
            :show => true,
         :default => nil,
     :perf_by_tag => nil
}
```

- update:
   Tag and classification of mapping is updated correctly.

- delete:
   Tags/Classifications/Taggings are deleted from resources (Vms..) when mapping is deleted.

### All Entities(All Providers,...) 
-  mapped tags/classifications are visible and usable in UI(expression editor, policy,... ) and mapped to existing category


#### Used Values in UI form:
- Entity: `All Entities`
- Resource Label: `environment`
- Tag Category:  `Department ` (this category have to exist)

Actions:
- create:
```
ContainerLabelTagMapping.last
=>

#<ContainerLabelTagMapping:0x00007fb33a935d60> {
                       :id => 12,
    :labeled_resource_type => "_all_entities_",
               :label_name => "environment",
              :label_value => nil,
                   :tag_id => 40,
               :created_at => Fri, 21 Aug 2020 07:53:26 UTC +00:00,
               :updated_at => Fri, 21 Aug 2020 07:53:26 UTC +00:00
}
```

```
# existing tag 
ContainerLabelTagMapping.last.tag 
=>

#<Tag:0x00007fb33b1bcf88> {
      :id => 40,
    :name => "/managed/department"
}
```

```
# existing classification 
ContainerLabelTagMapping.last.tag.classification
=>
#<Classification:0x00007fb33b4fb028> {
              :id => 40,
     :description => "Department",
            :icon => nil,
       :read_only => false,
          :syntax => "string",
    :single_value => false,
    :example_text => "The department the resource is assigned to, such as HR, Accounting, or Sales.",
          :tag_id => 40,
       :parent_id => nil,
            :show => true,
         :default => true,
     :perf_by_tag => nil
}
```

- update Tag Category:
 Tags/Classifications/Taggings **remain** to connected with previous Tag Category and tag:

 Mapping to Tag Category: `Department`
```
Vm.first.tags
=>

[
    [0] #<Tag:0x00007fb33b5d2640> {
          :id => 209,
        :name => "/managed/department/production"
    }
]
```
When Tag Category on the mapping is changed to the `Environment`:


```
Vm.first.tags
=>

[
    [0] #<Tag:0x00007fb33b5d2780> {
          :id => 208,
        :name => "/managed/environment/production"
    },
    [1] #<Tag:0x00007fb33b5d2640> {
          :id => 209,
        :name => "/managed/department/production".  ## <-- From Previous Tag Category
    }
]
```

- delete mapping:
 Tags/Classifications/Taggings created by mapping process **are not deleted** on resources.

### Links
Addresses https://github.com/ManageIQ/manageiq/issues/20409
Required Core PR - ManageIQ/manageiq#20289

lpichler commented:
Summarization of behaviour 

### Specific Entity (Provider,... ) - tags/classifications contain prefixes

#### Used Values in UI form:
- Entity: `Instance(Amazon)`
- Resource Label: `environment`
- Tag Category:  `NewCategory` (this category does not have to exist)

Actions:
- create:
  - Tag Mapping:
```
ContainerLabelTagMapping.last
=>

#<ContainerLabelTagMapping:0x00007fb3377cc1c0> {
                       :id => 9,
    :labeled_resource_type => "Vm",
               :label_name => "environment",
              :label_value => nil,
                   :tag_id => 215,
               :created_at => Thu, 20 Aug 2020 14:45:37 UTC +00:00,
               :updated_at => Thu, 20 Aug 2020 14:45:37 UTC +00:00
}
```
Tag created with mapping:

```
ContainerLabelTagMapping.last.tag
=>
#<Tag:0x00007fb31d1e4740> {
      :id => 215,
    :name => "/managed/amazon:vm:newcategory"
}
```
and classification

```
ContainerLabelTagMapping.last.tag.classification
=>
#<Classification:0x00007fb32e6d3330> {
              :id => 215,
     :description => "amazon:vm|NewCategory",
            :icon => nil,
       :read_only => true,
          :syntax => "string",
    :single_value => true,
    :example_text => nil,
          :tag_id => 215,
       :parent_id => nil,
            :show => true,
         :default => nil,
     :perf_by_tag => nil
}
```

- update:
   Tag and classification of mapping is updated correctly.

- delete:
   Tags/Classifications/Taggings are deleted from resources (Vms..) when mapping is deleted.

### All Entities(All Providers,...) 
-  mapped tags/classifications are visible and usable in UI(expression editor, policy,... ) and mapped to existing category


#### Used Values in UI form:
- Entity: `All Entities`
- Resource Label: `environment`
- Tag Category:  `Department ` (this category have to exist)

Actions:
- create:
```
ContainerLabelTagMapping.last
=>

#<ContainerLabelTagMapping:0x00007fb33a935d60> {
                       :id => 12,
    :labeled_resource_type => "_all_entities_",
               :label_name => "environment",
              :label_value => nil,
                   :tag_id => 40,
               :created_at => Fri, 21 Aug 2020 07:53:26 UTC +00:00,
               :updated_at => Fri, 21 Aug 2020 07:53:26 UTC +00:00
}
```

```
# existing tag 
ContainerLabelTagMapping.last.tag 
=>

#<Tag:0x00007fb33b1bcf88> {
      :id => 40,
    :name => "/managed/department"
}
```

```
# existing classification 
ContainerLabelTagMapping.last.tag.classification
=>
#<Classification:0x00007fb33b4fb028> {
              :id => 40,
     :description => "Department",
            :icon => nil,
       :read_only => false,
          :syntax => "string",
    :single_value => false,
    :example_text => "The department the resource is assigned to, such as HR, Accounting, or Sales.",
          :tag_id => 40,
       :parent_id => nil,
            :show => true,
         :default => true,
     :perf_by_tag => nil
}
```

- update Tag Category:
 Tags/Classifications/Taggings **remain** to connected with previous Tag Category and tag:

 Mapping to Tag Category: `Department`
```
Vm.first.tags
=>

[
    [0] #<Tag:0x00007fb33b5d2640> {
          :id => 209,
        :name => "/managed/department/production"
    }
]
```
When Tag Category on the mapping is changed to the `Environment`:


```
Vm.first.tags
=>

[
    [0] #<Tag:0x00007fb33b5d2780> {
          :id => 208,
        :name => "/managed/environment/production"
    },
    [1] #<Tag:0x00007fb33b5d2640> {
          :id => 209,
        :name => "/managed/department/production".  ## <-- From Previous Tag Category
    }
]
```

- delete mapping:
 Tags/Classifications/Taggings created by mapping process **are not deleted** on resources.

lpichler commented:
@gtanzillo 
this is rebased and squashed version of #7137 with [commit](https://github.com/ManageIQ/manageiq-ui-classic/commit/c13896b16bf3b6b34d5071301ed3b083a42a30d5) and in addition last commit fixes issue when tag is updated.

lpichler commented:
New warning message:

<img width="1042" alt="Screenshot 2020-09-11 at 14 28 03" src="https://user-images.githubusercontent.com/14937244/92925837-6a300080-f43b-11ea-8515-4f6ffc6e5f2c.png">

Fryguy commented:
For the warning, should we be using the flash-like warning like we have with Advanced Settings screen?

gtanzillo commented:
@Fryguy @agrare @lpichler Here are some notes for the demo:
- Since the user is required to enter an existing category when creating a mapping for `All Entities`, and the description is case sensitive, we should provide a drop-down list of all the existing categories for the user to select from.
  - Well add it to this PR
- Need to confirm whether a case sensitive lookup of mapping by label name is performed during refresh and if so, it should be done case-insensitively.
- Should log a warning when 2 mappings to the same single-value category are encountered resulting in only one of them being applied to the resource
- Should support mapping by provider type, but without prefixing the tag. Sort of a mix between the original and new behaviors
- Need to consider pluggability. Both the old and the new style
- We probably need an API endpoint for managing mappings

Fryguy commented:
> - Should support mapping by provider type, but without prefixing the tag. Sort of a mix between the original and new behaviors

This can be a follow up, since it would be net-new, and I'd prefer to iterate towards that solution.

lpichler commented:
I added changes from the DEMO meeting which were relevant for this "first phase::

- [x]  Since the user is required to enter an existing category when creating a mapping for All Entities, and the description is case sensitive, we should provide a drop-down list of all the existing categories for the user to select from.

- [x] Need to confirm whether a case sensitive lookup of mapping by label name is performed during refresh and if so, it should be done case-insensitively.
  - I am confirming that sensitive lookup of mapping by label nam was used here upon and I changed it to case insensitive versions. According some spec and research I found that this is related to provider's type and most providers have labels case sensitive:

      Provider | CASE | Link
      --- | --- | --- 
      Amazon |  Sensitive | https://docs.aws.amazon.com/general/latest/gr/aws_tagging.html
      Openstack |  Both? | https://docs.openstack.org/keystone/latest/admin/case-insensitive.html
      Azure |  Insensitive | https://docs.microsoft.com/en-us/rest/api/resources/tags/createorupdate
      Kubernetes + OpenShift |  Sensitive | Unable to find source(just from spec )

      - I am not sure how to determine it in Openstack but I added possibility to use case sensitivity/insensitivity according to the provider's type that we will rely on result provider's method `case_insensitivity_labels?`. 
     - default behaviour is case sensitivity because it is in more providers.
     - if we want to disable case insensitivity for lookup label name of mapping we have to define method on `case_insensitivity_labels?` on the provider and return `false`. Maybe we can use supports_feature? mechanism and currently we have to define the method for azure provider. I will do it once I get feedback on this. 

  - Done in those two commits:
    https://github.com/ManageIQ/manageiq/pull/20289/commits/c01324ebab80fc322dc04ea4461ec8974b338abb
    https://github.com/ManageIQ/manageiq/pull/20289/commits/c01324ebab80fc322dc04ea4461ec8974b338abb
- [x] Should log a warning when 2 mappings to the same single-value category are encountered resulting in only one of them being applied to the resource
  - done in core PR https://github.com/ManageIQ/manageiq/pull/20289/commits/efea44680dbf2d31a54c269b2037cf2c0c595321


Rest of point from demo meeting:

Points in [comment](https://github.com/ManageIQ/manageiq-ui-classic/pull/7273#issuecomment-694951389) would be added in follow up PR.


@gtanzillo @agrare @Fryguy Now PR is ready for review as suggestions from DEMO meeting were added ^^^^

gtanzillo commented on app/controllers/ops_controller/settings/label_tag_mapping.rb:
```diff
@@ -150,14 +161,32 @@ def lt_map_set_new_form_vars
       page << javascript_prologue
+
+      if params[:entity]
+        page.replace("tag_category_div", :partial => "ops/label_tag_mapping/tag_category",
```
@lpichler I'm trying to run this locally and getting this error for the missing partial -
```
[----] F, [2020-10-06T14:04:46.999003 #22945:3fed75120b90] FATAL -- : Error caught: [ActionView::MissingTemplate] Missing partial ops/label_tag_mapping/_tag_category with {:locale=>[:en], :formats=>[:html], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby, :haml, :coffee, :rjs]}. Searched in:
  * "/Users/greggtanzillo/work/ibm/manageiq-ui-classic/app/views"
  * "/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/bundler/gems/manageiq-v2v-6f8e502c5e2f/app/views"

/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/actionview-5.2.4.4/lib/action_view/path_set.rb:48:in `find'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/actionview-5.2.4.4/lib/action_view/lookup_context.rb:116:in `find'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/actionview-5.2.4.4/lib/action_view/renderer/partial_renderer.rb:423:in `find_template'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/actionview-5.2.4.4/lib/action_view/renderer/partial_renderer.rb:418:in `find_partial'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/actionview-5.2.4.4/lib/action_view/renderer/partial_renderer.rb:299:in `render'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/actionview-5.2.4.4/lib/action_view/renderer/renderer.rb:49:in `render_partial'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/actionview-5.2.4.4/lib/action_view/renderer/renderer.rb:23:in `render'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/actionview-5.2.4.4/lib/action_view/helpers/rendering_helper.rb:33:in `render'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/jquery-rjs-0.1.1.1/lib/action_view/helpers/jquery_helper.rb:566:in `block in render'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/jquery-rjs-0.1.1.1/lib/action_view/helpers/jquery_helper.rb:579:in `with_formats'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/jquery-rjs-0.1.1.1/lib/action_view/helpers/jquery_helper.rb:563:in `render'
/Users/greggtanzillo/.rvm/gems/ruby-2.5.3@vmdb_latest/gems/jquery-rjs-0.1.1.1/lib/action_view/helpers/jquery_helper.rb:381:in `replace'
/Users/greggtanzillo/work/ibm/manageiq-ui-classic/app/controllers/ops_controller/settings/label_tag_mapping.rb:186:in `block in label_tag_mapping_field_changed'
```
I can't seem to find it. Is it possible that it was not pushed up tho your branch?

lpichler replied:
@gtanzillo yes that's right I did not push haml file `ops/label_tag_mapping/tag_category. haml`
it is added now, thanks

gtanzillo commented:
@Fryguy @agrare Anything else needed for this PR before it can be merged?

agrare commented on app/controllers/ops_controller/settings/label_tag_mapping.rb:
```diff
@@ -12,12 +12,16 @@ module OpsController::Settings::LabelTagMapping
+    ALL_ENTITIES          => MappableEntity.new(nil,
+                                                "All Entities"),
     "Vm"                  => MappableEntity.new("amazon:vm:",
                                                 "ManageIQ::Providers::Amazon::CloudManager::Vm"),
```
Not related to this change at all, but this part doesn't look very pluggable, I'll add a task to our pluggability checklist

agrare commented on app/controllers/ops_controller/settings/label_tag_mapping.rb:
```diff
@@ -150,14 +161,32 @@ def lt_map_set_new_form_vars
+  def categories_for_select
+    categories = Classification.categories.sort_by(&:description)
+    categories.each_with_object([]) do |category, result|
+      next if category.read_only?
```
Looks like you could do `Classification.categories.reject(&:read_only?).sort_by(&:description)` above

agrare replied:
Actually if all you want is the description you can just `.map(&:description)` and this is a one-liner
